contract/registry/src/lib.rs

```
@@ -97,30 +95,80 @@ impl Contract {
     pub fn add_version(
         &mut self,
         #[serializer(borsh)] version_key: String,
+        #[serializer(borsh)] mode: DeployMode,
         #[serializer(borsh)] code: Vec<u8>,
-    ) {
-        assert_one_yocto();
+    ) -> PromiseOrValue<()> {
         self.assert_owner();
         require!(
             !self.versions.contains_key(&version_key),
             "Version key already exists",
         );
 
-        self.versions.insert(version_key, VersionEntry::new(code));
+        let hash = env::sha256_array(&code);
+
+        match mode {
+            DeployMode::Normal => {
+                assert_one_yocto();
+                let version_entry = VersionEntry::Code {
+                    hash,
+                    code: Some(code.clone()),
+                };
+                self.versions.insert(version_key, version_entry);
+                PromiseOrValue::Value(())
+            }
+            DeployMode::GlobalHash => {
+                let deposit = env::attached_deposit();
+                require!(
```

We should add a runtime config parameter for the minimum global contract deployment cost and fail early here. It is different between testnet <> mainnet.

---

I would like to, but I believe that:

- we cannot access those data from the context of a smart contract, meaning we would have to provide the value, either:
   - immutably during deployment, meaning the contract cannot respond to protocol configuration changes, or
   - the contract relies on a trusted oracle to update this value

I suppose we could go with one of these two options if we need to fail fast. What do you think is best here?

---

I believe that the registry is intended to be a templar-owned protocol singleton, if this is correct i would think it's okay to be a mutable configuration in the contract itself. It does not really impose any security risk.

---

I think this is best to put in another issue. #247